src/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -0,0 +1,194 @@
+                    source = new Span<byte>(_buffer, 0, readBytes);
+                }
+
+                lastResult = _decoder.Decompress(source, destination, out bytesConsumed, out bytesWritten);
+                if (lastResult == OperationStatus.InvalidData)
+                    throw new InvalidOperationException(SR.BrotliStream_Decompress_InvalidData);
+                if (bytesConsumed > 0)
+                    source = source.Slice(bytesConsumed);
+                if (bytesWritten > 0)
+                {
+                    totalWritten += bytesWritten;
+                    destination = destination.Slice(bytesWritten);
+                }
+            }
+
+            return totalWritten;
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState) =>
+            TaskToApm.Begin(ReadAsync(buffer, offset, count, CancellationToken.None), asyncCallback, asyncState);
+
+        public override int EndRead(IAsyncResult asyncResult) =>
+            TaskToApm.End<int>(asyncResult);
+
+        public override Task<int> ReadAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateParameters(array, offset, count);
+            return ReadAsyncMemory(new Memory<byte>(array, offset, count), cancellationToken).AsTask();
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (GetType() != typeof(BrotliStream))
+            {
+                // Ensure that existing streams derived from BrotliStream and that override ReadAsync(byte[],...)
+                // get their existing behaviors when the newer Memory-based overload is used.
+                return base.ReadAsync(destination, cancellationToken);
+            }
+            else
+            {
+                return ReadAsyncMemory(destination, cancellationToken);
+            }
+        }
+
+        internal ValueTask<int> ReadAsyncMemory(Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (_mode != CompressionMode.Decompress)
+                throw new InvalidOperationException(SR.BrotliStream_Compress_UnsupportedOperation);
+            EnsureNoActiveAsyncOperation();
+            EnsureNotDisposed();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
+            }
+            bool cleanup = true;
+            AsyncOperationStarting();
+            try
+            {
+                cleanup = false;
+                return FinishReadAsyncMemory(destination, cancellationToken);
+            }
+            finally
+            {
+                // if we haven't started any async work, decrement the counter to end the transaction
+                if (cleanup)
```

Comment:
In what situation will we ever be in the finally block with cleanup == true?  The only one would seem to be a thread abort, which doesn't exist in core.

Reply:
Thanks, you're correct that this is not necessary. ```cleanup``` is a holdover from some older code that performed operations within the try block before calling FinishReadAsyncMemory.